File: pkg/automerge/internal/native/encode_test.go
// Copyright (c) 2026 Probo Inc <hello@probo.com>.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package native

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestChangeEncodingExpandedRoundTrip reproduces the upstream
// test_change_encoding_expanded_change_round_trip: a change decoded from its
// canonical bytes re-encodes to exactly those bytes.
func TestChangeEncodingExpandedRoundTrip(t *testing.T) {
	t.Parallel()

	changeBytes := []byte{
		0x85, 0x6f, 0x4a, 0x83, // magic
		0xb2, 0x98, 0x9e, 0xa9, // checksum
		1, 61, 0, 2, 0x12, 0x34, // chunk type: change, length, deps, actor '1234'
		1, 1, 252, 250, 220, 255, 5, // seq, startOp, time
		14, 73, 110, 105, 116, 105, 97, 108, 105, 122, 97, 116, 105, 111, 110, // "Initialization"
		0, 6, // actor list, column count
		0x15, 3, 0x34, 1, 0x42, 2, // keyStr, insert, action
		0x56, 2, 0x57, 1, 0x70, 2, // valLen, valRaw, predNum
		0x7f, 1, 0x78, // keyStr: 'x'
		1,       // insert: false
		0x7f, 1, // action: set
		0x7f, 19, // valLen: 1 byte of type uint
		1,       // valRaw: 1
		0x7f, 0, // predNum: 0
		0, 1, 2, 3, 4, 5, 6, 7, 8, 9, // 10 trailing bytes inside the chunk
	}

	document, consumed, err := DecodeIncremental(changeBytes)
	require.NoError(t, err)
	require.Equal(t, len(changeBytes), consumed)
	require.Len(t, document.Changes, 1)

	encoded, err := EncodeChange(&document.Changes[0])
	require.NoError(t, err)
	assert.Equal(t, changeBytes, encoded)
}

func TestEncodeRLE_CanonicalRuns(t *testing.T) {
	t.Parallel()

	tests := map[string]struct {
		values []optional[uint64]
		want   []byte
	}{
		"all null": {
			values: []optional[uint64]{{}, {}},
			want:   nil,
		},
		"null then value": {
			values: []optional[uint64]{{}, {}, some(uint64(3))},
			want:   []byte{0, 2, 0x7f, 3},
		},
		"repeated values": {
			values: []optional[uint64]{some(uint64(0)), some(uint64(0))},
			want:   []byte{2, 0},
		},
		"literal then repeated then literal": {
			values: []optional[uint64]{
				some(uint64(1)),
				some(uint64(2)),
				some(uint64(2)),
				some(uint64(3)),
			},
			want: []byte{0x7f, 1, 2, 2, 0x7f, 3},
		},
		"literal values": {
			values: []optional[uint64]{
				some(uint64(1)),
				some(uint64(2)),
				some(uint64(3)),
			},
			want: []byte{0x7d, 1, 2, 3},
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			assert.Equal(t, test.want, encodeRLE(test.values, appendULEB))
		})
	}
}

File: packages/cookie-banner-tcf/src/encode.test.ts
// Copyright (c) 2026 Probo Inc <hello@probo.com>.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import { TCString } from "@iabtechlabtcf/core";
import type { BannerConfig, TCFGVL } from "@probo/cookie-banner";
import { describe, expect, it } from "vitest";

import { encodeTCString, gdprApplies, requireCmpVersion } from "./encode";

const vendorId = 52;
const unusedVendorId = 99;

const tcfGvl: TCFGVL = {
  gvlSpecificationVersion: 3,
  vendorListVersion: 42,
  tcfPolicyVersion: 5,
  lastUpdated: "2026-01-15T17:00:00Z",
  purposes: {
    "1": {
      id: 1,
      name: "Store and/or access information on a device",
      description: "Cookies, device or similar online identifiers.",
    },
    "2": {
      id: 2,
      name: "Select basic ads",
      description: "Ads can be shown based on limited data.",
    },
  },
  specialPurposes: {},
  features: {},
  specialFeatures: {
    "1": {
      id: 1,
      name: "Use precise geolocation data",
      description: "Your precise geolocation data can be used.",
    },
  },
  stacks: {},
  vendors: {
    [String(vendorId)]: {
      id: vendorId,
      name: "Test Vendor",
      purposes: [1],
      legIntPurposes: [],
      flexiblePurposes: [],
      specialPurposes: [],
      features: [],
      specialFeatures: [1],
      policyUrl: "https://example.com/privacy",
      usesCookies: true,
      cookieMaxAgeSeconds: 86400,
      cookieRefresh: false,
      usesNonCookieAccess: false,
    },
    [String(unusedVendorId)]: {
      id: unusedVendorId,
      name: "Unused Vendor",
      purposes: [2],
      legIntPurposes: [],
      flexiblePurposes: [],
      specialPurposes: [],
      features: [],
      specialFeatures: [],
      policyUrl: "https://unused.example.com/privacy",
    },
  },
};

function bannerConfig(overrides: Partial<BannerConfig> = {}): BannerConfig {
  return {
    banner_id: "banner",
    version: 1,
    language: "en",
    default_language: "en",
    cookie_policy_url: "https://example.com/cookies",
    consent_expiry_days: 365,
    consent_mode: "OPT_IN",
    regulation: "GDPR",
    layout: {
      presentation: "OPT_IN",
      initial_state: "banner",
      reopen_state: "panel",
      default_non_necessary_granted: false,
      buttons: {
        accept_all: true,
        reject_all: true,
        customize: true,
        save: true,
      },
      settings_link: "default",
    },
    show_branding: false,
    resource_reporting_enabled: false,
    tcf: {
      gvl: tcfGvl,
      cmp_id: 4095,
      cmp_version: 1,
      publisher_cc: "AA",
      policy_version: 5,
    },
    categories: [],
    texts: {},
    ...overrides,
  };
}

describe("gdprApplies", () => {
  it("is true for GDPR and UK_GDPR", () => {
    expect(gdprApplies(bannerConfig({ regulation: "GDPR" }))).toBe(true);
    expect(gdprApplies(bannerConfig({ regulation: "UK_GDPR" }))).toBe(true);
  });

  it("is false for other regulations", () => {
    expect(gdprApplies(bannerConfig({ regulation: "CCPA" }))).toBe(false);
    expect(gdprApplies(bannerConfig({ regulation: null }))).toBe(false);
  });
});

describe("encodeTCString", () => {
  it("throws when tcf.gvl is missing", () => {
    expect(() => encodeTCString(bannerConfig({ tcf: {} }), "all")).toThrow(
      /tcf.gvl is missing/,
    );
  });

  it("throws when tcf.cmp_id is missing", () => {
    expect(() => encodeTCString(bannerConfig({ tcf: { gvl: tcfGvl } }), "all")).toThrow(
      /tcf.cmp_id is missing or invalid/,
    );
  });

  it("throws when tcf.cmp_version is missing", () => {
    expect(() =>
      encodeTCString(bannerConfig({ tcf: { gvl: tcfGvl, cmp_id: 4095 } }), "all"),
    ).toThrow(/tcf.cmp_version is missing or invalid/);
  });

  it("throws when tcf.cmp_version exceeds the 12-bit field", () => {
    expect(() => requireCmpVersion(4096)).toThrow(/tcf.cmp_version is missing or invalid/);
  });

  it("encodes disclosed vendors on reject without granting consent", () => {
    const encoded = encodeTCString(bannerConfig(), "none");
    const decoded = TCString.decode(encoded);

    expect(decoded.cmpId).toBe(4095);
    expect(decoded.cmpVersion).toBe(1);
    expect(decoded.isServiceSpecific).toBe(true);
    expect(decoded.vendorsDisclosed.has(vendorId)).toBe(true);
    expect(decoded.vendorConsents.has(vendorId)).toBe(false);
    expect(decoded.purposeConsents.has(1)).toBe(false);
    expect(decoded.specialFeatureOptins.has(1)).toBe(false);
  });

  it("encodes vendor, purpose, and special-feature grants on accept", () => {
    const encoded = encodeTCString(bannerConfig(), "all");
    const decoded = TCString.decode(encoded);

    expect(decoded.vendorsDisclosed.has(vendorId)).toBe(true);
    expect(decoded.vendorConsents.has(vendorId)).toBe(true);
    expect(decoded.purposeConsents.has(1)).toBe(true);
    expect(decoded.specialFeatureOptins.has(1)).toBe(true);
  });

  it("encodes only the selected purpose, vendor, and special-feature bits", () => {
    const encoded = encodeTCString(bannerConfig(), {
      purposeConsents: [1],
      purposeLegitimateInterests: [],
      vendorConsents: [vendorId],
      vendorLegitimateInterests: [],
      specialFeatureOptins: [],
    });
    const decoded = TCString.decode(encoded);

    expect(decoded.purposeConsents.has(1)).toBe(true);
    expect(decoded.purposeConsents.has(2)).toBe(false);
    expect(decoded.purposeLegitimateInterests.has(1)).toBe(false);
    expect(decoded.vendorConsents.has(vendorId)).toBe(true);
    expect(decoded.vendorConsents.has(unusedVendorId)).toBe(false);
    expect(decoded.vendorLegitimateInterests.has(vendorId)).toBe(false);
    expect(decoded.specialFeatureOptins.has(1)).toBe(false);
  });
});
